src/lib/work_selector/work_lib.ml

```diff
@@ -37,7 +27,7 @@ module Make (Inputs : Intf.Inputs_intf) = struct
                 whenever the pipe has broadcasted new frontier. The works
                 between consecutive frontier broadcasts should be largely
                 identical. *)
-      ; mutable jobs_seen : Job_status.t Seen_key.Map.t
+      ; jobs_scheduled : Job_key.t Hash_set.t
```

Comment:
Please no hashes here.

We used to have a hash map here, and it had a terrible performance.

Computing a hash for the statement is much more expensive than doing a comparison (because comparison would likely short-circuit on the first field of the structure, while hash has to traverse the entire structure).

Reply:
Let's add a doc-comment.

There is a semantical change: now these are jobs that are scheduled or processed.

We do clean-up processed jobs only once they appear in a block. Previously we would do it immediately upon completion.

I don't think the change is substantial, because we always call `get_expensive_work` on top of the `all_unscheduled_works`, hence we were filtering out our own works via this mechanism. Current implementation does a similar thing, just that filtering of our own completed work would happen in  `all_unscheduled_works` rather than in `get_expensive_work`.

Reply:
I'd suggest to merge `all_unseen_works` and `get_expensive_work` in a follow-up PR, it makes no sense to keep these two separate. If we had these as one function, it would be clearer in this PR that the logic of `all_unseen_works + get_expensive_work` before the PR is the same as `all_unscheduled_works + get_expensive_work` after.